// ngx-autofocus-fix/esm5/lib/autofocus-fix.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken, NgModule, Optional } from '@angular/core';
import { AutofocusFixConfig } from './autofocus-fix-config';
import { AutofocusFixDirective } from './autofocus-fix.directive';
import { noAutofocusFixConfigError } from './no-autofocus-fix-config.error';
// Exists for AoT support
/**
 * @param {?} options
 * @return {?}
 */
export function configFactory(options) {
    return new AutofocusFixConfig(options);
}
// Exists for AoT support
/** @type {?} */
var AutofocusFixOptionsInternalToken = new InjectionToken('AutofocusFixOptions');
var AutofocusFixModule = /** @class */ (function () {
    function AutofocusFixModule($config) {
        if (!$config) {
            noAutofocusFixConfigError();
        }
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    AutofocusFixModule.forRoot = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        if (options === void 0) { options = {}; }
        return {
            ngModule: AutofocusFixModule,
            providers: [
                {
                    provide: AutofocusFixOptionsInternalToken,
                    useValue: options,
                },
                {
                    provide: AutofocusFixConfig,
                    useFactory: configFactory,
                    deps: [AutofocusFixOptionsInternalToken],
                },
            ],
        };
    };
    AutofocusFixModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [AutofocusFixDirective],
                    exports: [AutofocusFixDirective]
                },] }
    ];
    /** @nocollapse */
    AutofocusFixModule.ctorParameters = function () { return [
        { type: AutofocusFixConfig, decorators: [{ type: Optional }] }
    ]; };
    return AutofocusFixModule;
}());
export { AutofocusFixModule };
