// ngx-autofocus-fix/esm2015/lib/autofocus-fix.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken, NgModule, Optional } from '@angular/core';
import { AutofocusFixConfig } from './autofocus-fix-config';
import { AutofocusFixDirective } from './autofocus-fix.directive';
import { noAutofocusFixConfigError } from './no-autofocus-fix-config.error';
// Exists for AoT support
/**
 * @param {?} options
 * @return {?}
 */
export function configFactory(options) {
    return new AutofocusFixConfig(options);
}
// Exists for AoT support
/** @type {?} */
const AutofocusFixOptionsInternalToken = new InjectionToken('AutofocusFixOptions');
export class AutofocusFixModule {
    /**
     * @param {?} $config
     */
    constructor($config) {
        if (!$config) {
            noAutofocusFixConfigError();
        }
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    static forRoot(options = {}) {
        return {
            ngModule: AutofocusFixModule,
            providers: [
                {
                    provide: AutofocusFixOptionsInternalToken,
                    useValue: options,
                },
                {
                    provide: AutofocusFixConfig,
                    useFactory: configFactory,
                    deps: [AutofocusFixOptionsInternalToken],
                },
            ],
        };
    }
}
AutofocusFixModule.decorators = [
    { type: NgModule, args: [{
                declarations: [AutofocusFixDirective],
                exports: [AutofocusFixDirective]
            },] }
];
/** @nocollapse */
AutofocusFixModule.ctorParameters = () => [
    { type: AutofocusFixConfig, decorators: [{ type: Optional }] }
];
